### DISCUSSION TITLE
Counting how much mass is added when adding a density floor

### DISCUSSION THREAD

**Discussion by ThomasJannaudCAM** [idefix-code/idefix#272] [General]
Hi everyone,

I am currently running simulations with dust, adding both a density and a velocity floor, in the InternalBoundary of the setup.cpp
I would like to have an idea of what exactly these floors are doing over the course of the simulation. For instance with the dust density floor, I would like to know how much dust mass I am adding and where. One way to do it would be by an user defined variable. What I would do is create an IdefixArray3D called "added_dust", and in the if condition of the density floor, add the density I am adding to dust->Vc(RHO,k,j,i) to the (k,j,i) cell of "added_dust". Then in ComputeUserVars call "added_dust" so that I can enroll it later in the Setup::Setup.
I don't think I can use the analysis.cpp/hpp for this, because those are called after InternalBoundary, thus after all the cells in which the density is under the floor have been modified. But I might be wrong.
Has any of you thought about this already ? Would you know how to do it in an optimized way ?

Many thanks in advance.

**Comment by ThomasJannaudCAM** [comment]
Quick update: That solution appears to work properly.

What I do is at the very start of the setup.cpp (outside any loop) define a global `IdefixHostArray3D<real> DustAdded`
I call it at the start of the `InternalBoundary`, giving it the dimensions `data.np_tot[IDIR],data.np_tot[JDIR],data.np_tot[KDIR]`
Inside the `idefix_for` loop of the `InternalBoundary`, there is the following if condition:
```
if(Vcd(RHO,k,j,i)<floor){
DustAdded(k,j,i)+=floor-Vcd(RHO,k,j,i);
Vcd(RHO,k,j,i)=floor;
}
```
Finally in the `ComputeUserVars` I add the following:
```
IdefixHostArray3D<real> AddedDust  = variables["DustAdd"];
AddedDust = DustAdded;
```
where `DustAdd` is an user-defined variable.

However this solution is rather ugly, and I am open to ideas that don't rely on creating a global 3D array.

Thanks !

**Reply by glesur** [reply]
Hi @ThomasJannaudCAM , I think your solution (stacking the added density into an array allocated in the Setup constructor) is the way to go. You'll have to define this array globally as I believe you did. The "best" way to do that is to define a class wrapping the array, as in the [example described the documentation](https://idefix.readthedocs.io/latest/reference/setup.cpp.html#i-need-a-global-idefixarray-for-my-setup).

One last thing: you appear to be using a IdefixHostArray in an idefix_for. This won't be portable. Best is to use a IdefixArray for the idefix_for, and then a `Kokkos::deep_copy(DustAdded, variables["DustAdd"])` in ComputeUserVars.

**Reply by ThomasJannaudCAM** [reply]
Hi @glesur, thanks a lot ! I had missed this part of the documentation.

**Comment by ThomasJannaudCAM** [comment]
Hi !
Sorry for bothering you but I'm coming back to this, as I'm trying to make it work for multiple dust species.
Is the "best" way to do this to simply change all 3D idefixarrays to 4D ones ? Like the following

```
class MyGlobalClass {
public:
   //Class constructor
   MyGlobalClass(DataBlock &data) {
   //allocate some memory for the array the class contains
      this->DustAdded = IdefixArray4D<real>("DustAdded",data.dust.size(),data.np_tot[KDIR], data.np_tot[JDIR], data.np_tot[IDIR]);
}

   //DustAdded, member of the class
   IdefixArray4D<real> DustAdded;
};
```

and adding a loop on `data.dust.size()` in the `InternalBoundary` to account for all species.

Or is there something else that would be better optimized ?

**Reply by ThomasJannaudCAM** [reply]
Oh it seems user variables have to be 3D arrays, I'll find something else :)
